### PR TITLE
Reimplement Macros with Scala 3 metaprogramming

### PR DESCRIPTION
**Pull request by cchantep**
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [ ] Have you [squashed your commits](https://www.playframework.com/documentation/latest/WorkingWithGit#Squashing-commits)?
* [x] Have you added copyright headers to new files?
* [x] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?

## Purpose

Reimplement Macros with Scala 3 metaprogramming.

- Restore support of non case class (previously thanks to custom extractor/unapply, now using `ProductOf`/`Conversion`)
- Restore support of non case object
- Restore Value class support
- Restore custom naming support (field & discriminator)

## Background Context

First, supports same compile-time features as those provided by the Scala 2 macros.

Simple Dotty derivation not sufficient, use new meta-programming to wrap/adapt (mirror still used to resolve the class fields).

## References

Are there any relevant issues / PRs / mailing lists discussions?

### CODE REVIEW

File: play-json/shared/src/test/scala-2/play/api/libs/json/MacroScala2Spec.scala [hunk 1]
```diff
@@ -12,179 +12,8 @@ import org.scalacheck.Gen
 class MacroScala2Spec extends AnyWordSpec with Matchers with org.scalatestplus.scalacheck.ScalaCheckPropertyChecks {
   import MacroScala2Spec._
 
-  "Reads" should {
```
Comment (cchantep) [inline play-json/shared/src/test/scala-2/play/api/libs/json/MacroScala2Spec.scala:15]: Merge back to `MacroSpec`

File: play-json/shared/src/test/scala/play/api/libs/json/JsonSharedSpec.scala [hunk 2]
```diff
@@ -33,36 +33,34 @@ class JsonSharedSpec extends AnyWordSpec with Matchers with org.scalatestplus.sc
   "JSON" should {
     "equals JsObject independently of field order" in json { js =>
       js.obj(
-          "field1" -> 123,
```
Comment (cchantep) [inline play-json/shared/src/test/scala/play/api/libs/json/JsonSharedSpec.scala:36]: Source formatting

File: play-json/shared/src/test/scala/play/api/libs/json/MacroSpec.scala [hunk 3]
```diff
@@ -9,6 +9,7 @@ import Matchers._
 import org.scalatest.wordspec.AnyWordSpec
 
 object TestFormats {
+
```
Comment (cchantep) [inline play-json/shared/src/test/scala/play/api/libs/json/MacroSpec.scala:12]: Source formatting + restore common specs for Scala 2/3

File: project/Common.scala [hunk 4]
```diff
@@ -11,11 +11,11 @@ object Common extends AutoPlugin {
 
   override def globalSettings =
     Seq(
-      organization := "com.typesafe.play",
```
Comment (cchantep) [inline project/Common.scala:14]: Source formatting

File: .scalafmt.conf [hunk 5]
```diff
@@ -1,23 +1,54 @@
...
+literals.scientific=Upper
+
+fileOverride {
+  "glob:**/*/src/*/scala-3/**/*.scala" {
+    runner.dialect = scala3
+  }
+}
+
+project.excludePaths = [
+  "glob:**/docs/**/code*/*.scala"
```
Comment (cchantep) [inline .scalafmt.conf:53]: Make sure special comment to extract snipped are unchanged (special comment a doc delimiters)

File: play-json/shared/src/test/scala-3/play/api/libs/json/MacroScala3Spec.scala [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  "Case class" should {
+    "not be handled" when {
+      "no Product Conversion" in {
+        import MacroSpec.UsingAlias
+
+        "Macros.writer[UsingAlias]".mustNot(typeCheck)
+      }
+
+      "no custom ProductOf" in {
+        "Macros.writer[CustomNoProductOf]".mustNot(typeCheck)
```
Comment (cchantep) [inline play-json/shared/src/test/scala-3/play/api/libs/json/MacroScala3Spec.scala:23]: `Conversion` alone is not sufficient

**Comment by AlexITC** [comment]
Are there any plans to consider merging this anytime soon? @playframework/contributors

File: docs/manual/working/scalaGuide/main/json/code/ScalaJsonCombinatorsSpec.scala [hunk 7]
```diff
@@ -7,6 +7,7 @@ package scalaguide.json
 import org.specs2.mutable.Specification
 
 class ScalaJsonCombinatorsSpec extends Specification {
+
```
[the remark this answers is not in the document]
Comment (cchantep) [reply docs/manual/working/scalaGuide/main/json/code/ScalaJsonCombinatorsSpec.scala]: See https://github.com/playframework/play-json/pull/699

**Comment by BillyAutrey** [comment]
> Are there any plans to consider merging this anytime soon? @playframework/contributors 

Just for transparency here, we asked that Cédric split up the PR.  Separating the large amount of formatting changes from the relevant code makes it easier to review.

File: build.sbt [hunk 8]
```diff
@@ -85,6 +80,17 @@ val scalacOpts = Seq(
   "-Ywarn-macros:after"
 )
 
+val silencerVersion = "1.7.8"
```
Comment (ihostage) [inline build.sbt:83]: Silencer dropped in https://github.com/playframework/play-json/pull/695.

File: .github/workflows/build-test.yml [hunk 9]
```diff
@@ -81,7 +81,7 @@ jobs:
     strategy:
       matrix:
         jdk: [ 11, 8 ]
-        scala: [ 2.12.15, 2.13.8, 3.0.2 ] # Should be sync with Mergify conditions (.mergify.yml)
+        scala: [ 2.12.15, 2.13.7, 3.1.2-RC1-bin-20211222-c94b333-NIGHTLY ]
```
Comment (ihostage) [inline .github/workflows/build-test.yml:84]: Cedric, can you change only Scala 3 version without removing the comment and changing the `2.13.X` version?

Comment (cchantep) [reply .github/workflows/build-test.yml]: Could thanks the long time to revert, and so the conflicts introduced meanwhile ...

**Comment by ihostage** [comment]
@cchantep I think we are ready to merge that and just should wait for a stable release of Scala `3.1.2`.
But If somebody can review the code of macros, would be great, because I don't have the needed skills for that and can't do that. 😞

File: build.sbt [hunk 10]
```diff
@@ -10,23 +10,18 @@ import com.typesafe.tools.mima.core._
 import sbtcrossproject.CrossPlugin.autoImport.crossProject
 import sbtcrossproject.CrossType
 
+ThisBuild / scalafmtOnCompile := true
```
Comment (AlexITC) [inline build.sbt:13]: This option is discouraged by scalafmt itself

File: play-json/shared/src/main/scala-3/play/api/libs/json/ImplicitResolver.scala [hunk 11]
```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (C) 2009-2021 Lightbend Inc. <https://www.lightbend.com>
```
Comment (AlexITC) [inline play-json/shared/src/main/scala-3/play/api/libs/json/ImplicitResolver.scala:2]: I think this shouldn't be used anymore, please verify with someone that is more aware of these details

File: play-json/shared/src/test/scala-3/play/api/libs/json/MacroScala3Spec.scala [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+package play.api.libs.json
+
+import org.scalatest.matchers.must.Matchers
+import org.scalatest.wordspec.AnyWordSpec
+
+final class MacroScala3Spec
+    extends AnyWordSpec
+    with Matchers
+    with org.scalatestplus.scalacheck.ScalaCheckPropertyChecks {
+  "Case class" should {
```
Comment (AlexITC) [inline play-json/shared/src/test/scala-3/play/api/libs/json/MacroScala3Spec.scala:14]: Should we include success cases?

Comment (cchantep) [reply play-json/shared/src/test/scala-3/play/api/libs/json/MacroScala3Spec.scala:14]: Already covered in common `MacroSpec`

File: play-json/shared/src/main/scala-3/play/api/libs/json/ImplicitResolver.scala (hunk 11 above)
Comment (cchantep) [reply play-json/shared/src/main/scala-3/play/api/libs/json/ImplicitResolver.scala:2]: All the existing files have this, including the Scala2 version of `ImplicitResolver` on which this is based. Until it's removed everywhere, better to have it there.

**Comment by cchantep** [comment]
Up

**Comment by cchantep** [comment]
up

**Comment by cchantep** [comment]
Up

**Comment by cchantep** [comment]
Up

**Comment by mkurz** [comment]
Please fix the conflicts ,thanks!

**Comment by cchantep** [comment]
> Please fix the conflicts ,thanks!

Approval required

**Comment by godenji** [comment]
Thanks for pushing this forward @cchantep 

After this is merged (hopefully soon!), when can we expect a new release? I've been working off of the 2.10.0-RC5 release in a Scala 3 project; it's 100+ commits behind `main` at this point...

**Comment by mkurz** [comment]
OK, I am going to merge this and release 2.10.0-RC6 so people can start testing

**Comment by cchantep** [comment]
It would have been nice that I was able to merge it myself (as core contributor for "quite sometime") ...

**Comment by mkurz** [comment]
> It would have been nice that I was able to merge

You can do now.

**Comment by xuwei-k** [comment]
- https://github.com/playframework/play-json/issues/728

😭

**Pull request merged**